File: modules/sqlite-jdbc/src/main/java/io/roastedroot/sqlite4j/SQLiteConnectionConfig.java
package io.roastedroot.sqlite4j;

import static io.roastedroot.sqlite4j.SQLiteConfig.DEFAULT_DATE_STRING_FORMAT;

import io.roastedroot.sqlite4j.date.FastDateFormat;
import java.sql.Connection;
import java.util.EnumMap;
import java.util.Map;
import java.util.Objects;
import java.util.Properties;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

/** Connection local configurations */
public class SQLiteConnectionConfig implements Cloneable {
    private SQLiteConfig.@NonNull DateClass dateClass = SQLiteConfig.DateClass.INTEGER;
    private SQLiteConfig.@NonNull DatePrecision datePrecision =
            SQLiteConfig.DatePrecision.MILLISECONDS; // Calendar.SECOND or Calendar.MILLISECOND
    private @NonNull String dateStringFormat = DEFAULT_DATE_STRING_FORMAT;
    private @NonNull FastDateFormat dateFormat = FastDateFormat.getInstance(dateStringFormat);

    private int transactionIsolation = Connection.TRANSACTION_SERIALIZABLE;
    private SQLiteConfig.@NonNull TransactionMode transactionMode =
            SQLiteConfig.TransactionMode.DEFERRED;
    private boolean autoCommit = true;
    private boolean getGeneratedKeys = true;

    public static SQLiteConnectionConfig fromPragmaTable(@NonNull Properties pragmaTable) {
        return new SQLiteConnectionConfig(
                SQLiteConfig.DateClass.getDateClass(
                        pragmaTable.getProperty(
                                SQLiteConfig.Pragma.DATE_CLASS.pragmaName,
                                SQLiteConfig.DateClass.INTEGER.name())),
                SQLiteConfig.DatePrecision.getPrecision(
                        pragmaTable.getProperty(
                                SQLiteConfig.Pragma.DATE_PRECISION.pragmaName,
                                SQLiteConfig.DatePrecision.MILLISECONDS.name())),
                pragmaTable.getProperty(
                        SQLiteConfig.Pragma.DATE_STRING_FORMAT.pragmaName,
                        DEFAULT_DATE_STRING_FORMAT),
                Connection.TRANSACTION_SERIALIZABLE,
                SQLiteConfig.TransactionMode.getMode(
                        pragmaTable.getProperty(
                                SQLiteConfig.Pragma.TRANSACTION_MODE.pragmaName,
                                SQLiteConfig.TransactionMode.DEFERRED.name())),
                true,
                Boolean.parseBoolean(
                        pragmaTable.getProperty(
                                SQLiteConfig.Pragma.JDBC_GET_GENERATED_KEYS.pragmaName, "true")));
    }

    public SQLiteConnectionConfig(
            SQLiteConfig.@NonNull DateClass dateClass,
            SQLiteConfig.@NonNull DatePrecision datePrecision,
            @NonNull String dateStringFormat,
            int transactionIsolation,
            SQLiteConfig.@NonNull TransactionMode transactionMode,
            boolean autoCommit,
            boolean getGeneratedKeys) {
        setDateClass(dateClass);
        setDatePrecision(datePrecision);
        setDateStringFormat(dateStringFormat);
        setTransactionIsolation(transactionIsolation);
        setTransactionMode(transactionMode);
        setAutoCommit(autoCommit);
        setGetGeneratedKeys(getGeneratedKeys);
    }

    public SQLiteConnectionConfig copyConfig() {
        return new SQLiteConnectionConfig(
                dateClass,
                datePrecision,
                dateStringFormat,
                transactionIsolation,
                transactionMode,
                autoCommit,
                getGeneratedKeys);
    }

    public long getDateMultiplier() {
        return (datePrecision == SQLiteConfig.DatePrecision.MILLISECONDS) ? 1L : 1000L;
    }

    public SQLiteConfig.DateClass getDateClass() {
        return dateClass;
    }

    public void setDateClass(SQLiteConfig.@NonNull DateClass dateClass) {
        this.dateClass = dateClass;
    }

    public SQLiteConfig.DatePrecision getDatePrecision() {
        return datePrecision;
    }

    public void setDatePrecision(SQLiteConfig.@NonNull DatePrecision datePrecision) {
        this.datePrecision = datePrecision;
    }

    public String getDateStringFormat() {
        return dateStringFormat;
    }

    public void setDateStringFormat(@NonNull String dateStringFormat) {
        this.dateStringFormat = dateStringFormat;
        this.dateFormat = FastDateFormat.getInstance(dateStringFormat);
    }

    public FastDateFormat getDateFormat() {
        return dateFormat;
    }

    public boolean isAutoCommit() {
        return autoCommit;
    }

    public void setAutoCommit(boolean autoCommit) {
        this.autoCommit = autoCommit;
    }

    public int getTransactionIsolation() {
        return transactionIsolation;
    }

    public void setTransactionIsolation(int transactionIsolation) {
        this.transactionIsolation = transactionIsolation;
    }

    public SQLiteConfig.TransactionMode getTransactionMode() {
        return transactionMode;
    }

    @SuppressWarnings("deprecation")
    public void setTransactionMode(SQLiteConfig.@NonNull TransactionMode transactionMode) {
        this.transactionMode = transactionMode;
    }

    public boolean isGetGeneratedKeys() {
        return getGeneratedKeys;
    }

    public void setGetGeneratedKeys(boolean getGeneratedKeys) {
        this.getGeneratedKeys = getGeneratedKeys;
    }

    private static final Map<SQLiteConfig.TransactionMode, String> beginCommandMap =
            new EnumMap<>(SQLiteConfig.TransactionMode.class);

    static {
        beginCommandMap.put(SQLiteConfig.TransactionMode.DEFERRED, "begin;");
        beginCommandMap.put(SQLiteConfig.TransactionMode.IMMEDIATE, "begin immediate;");
        beginCommandMap.put(SQLiteConfig.TransactionMode.EXCLUSIVE, "begin exclusive;");
    }

    @Nullable String transactionPrefix() {
        return beginCommandMap.get(transactionMode);
    }

    @Override
    public boolean equals(@Nullable Object o) {
        if (this == o) return true;
        if (!(o instanceof SQLiteConnectionConfig)) return false;
        SQLiteConnectionConfig that = (SQLiteConnectionConfig) o;
        return transactionIsolation == that.transactionIsolation
                && autoCommit == that.autoCommit
                && getGeneratedKeys == that.getGeneratedKeys
                && dateClass == that.dateClass
                && datePrecision == that.datePrecision
                && Objects.equals(dateStringFormat, that.dateStringFormat)
                && Objects.equals(dateFormat, that.dateFormat)
                && transactionMode == that.transactionMode;
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                dateClass,
                datePrecision,
                dateStringFormat,
                dateFormat,
                transactionIsolation,
                transactionMode,
                autoCommit,
                getGeneratedKeys);
    }
}

File: modules/sqlite-jdbc/src/main/java/io/roastedroot/sqlite4j/SQLiteConnectionConfig.kt
package io.roastedroot.sqlite4j

import io.roastedroot.sqlite4j.date.FastDateFormat
import java.sql.Connection
import java.util.EnumMap
import java.util.Objects
import java.util.Properties

/** Connection local configurations */
class SQLiteConnectionConfig(
    var dateClass: SQLiteConfig.DateClass,
    var datePrecision: SQLiteConfig.DatePrecision,
    dateStringFormat: String,
    var transactionIsolation: Int,
    var transactionMode: SQLiteConfig.TransactionMode,
    var isAutoCommit: Boolean,
    var isGetGeneratedKeys: Boolean
) : Cloneable {
    var dateStringFormat: String = dateStringFormat
        set(value) {
            field = value
            dateFormat = FastDateFormat.getInstance(value)
        }

    var dateFormat: FastDateFormat = FastDateFormat.getInstance(dateStringFormat)
        private set

    fun copyConfig(): SQLiteConnectionConfig {
        return SQLiteConnectionConfig(
            dateClass,
            datePrecision,
            dateStringFormat,
            transactionIsolation,
            transactionMode,
            isAutoCommit,
            isGetGeneratedKeys
        )
    }

    val dateMultiplier: Long
        get() = if (datePrecision == SQLiteConfig.DatePrecision.MILLISECONDS) 1L else 1000L

    fun transactionPrefix(): String? {
        return beginCommandMap[transactionMode]
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is SQLiteConnectionConfig) return false
        return transactionIsolation == o.transactionIsolation &&
            isAutoCommit == o.isAutoCommit &&
            isGetGeneratedKeys == o.isGetGeneratedKeys &&
            dateClass == o.dateClass &&
            datePrecision == o.datePrecision &&
            Objects.equals(dateStringFormat, o.dateStringFormat) &&
            Objects.equals(dateFormat, o.dateFormat) &&
            transactionMode == o.transactionMode
    }

    override fun hashCode(): Int {
        return Objects.hash(
            dateClass,
            datePrecision,
            dateStringFormat,
            dateFormat,
            transactionIsolation,
            transactionMode,
            isAutoCommit,
            isGetGeneratedKeys
        )
    }

    companion object {
        private val beginCommandMap: MutableMap<SQLiteConfig.TransactionMode, String> =
            EnumMap(SQLiteConfig.TransactionMode::class.java)

        init {
            beginCommandMap[SQLiteConfig.TransactionMode.DEFERRED] = "begin;"
            beginCommandMap[SQLiteConfig.TransactionMode.IMMEDIATE] = "begin immediate;"
            beginCommandMap[SQLiteConfig.TransactionMode.EXCLUSIVE] = "begin exclusive;"
        }

        @JvmStatic
        fun fromPragmaTable(pragmaTable: Properties): SQLiteConnectionConfig {
            return SQLiteConnectionConfig(
                SQLiteConfig.DateClass.getDateClass(
                    pragmaTable.getProperty(
                        SQLiteConfig.Pragma.DATE_CLASS.pragmaName,
                        SQLiteConfig.DateClass.INTEGER.name
                    )
                ),
                SQLiteConfig.DatePrecision.getPrecision(
                    pragmaTable.getProperty(
                        SQLiteConfig.Pragma.DATE_PRECISION.pragmaName,
                        SQLiteConfig.DatePrecision.MILLISECONDS.name
                    )
                ),
                pragmaTable.getProperty(
                    SQLiteConfig.Pragma.DATE_STRING_FORMAT.pragmaName,
                    SQLiteConfig.DEFAULT_DATE_STRING_FORMAT
                ),
                Connection.TRANSACTION_SERIALIZABLE,
                SQLiteConfig.TransactionMode.getMode(
                    pragmaTable.getProperty(
                        SQLiteConfig.Pragma.TRANSACTION_MODE.pragmaName,
                        SQLiteConfig.TransactionMode.DEFERRED.name
                    )
                ),
                true,
                java.lang.Boolean.parseBoolean(
                    pragmaTable.getProperty(
                        SQLiteConfig.Pragma.JDBC_GET_GENERATED_KEYS.pragmaName, "true"
                    )
                )
            )
        }
    }
}
